### ISSUE TITLE
Arabic Language Support ?

### ISSUE THREAD

**Issue by ahmedhsilverkey** [QuestPDF/QuestPDF#94]
**Is your feature request related to a problem? Please describe.**
I am creating a report that displays some words in Arabic however the Arabic words are not being displayed properly

**Describe the solution you'd like**
Arabic (or languages that are not English) support would be great

**Describe alternatives you've considered**
I will be using https://github.com/webgio/Rotativa.AspNetCore for my reports instead for now

**Additional context**
this is a screenshot of some Arabic words in the report, it's shouldn't look like that
![image](https://user-images.githubusercontent.com/85054037/148678221-07bc5e72-d4b9-45d7-b4cb-8dc1a8c9e707.png)

**Comment by MarcinZiabek** [comment]
Hello 😁

Unfortunately, you are right. Currently, QuestPDF offers limited support for languages that require more advanced text shaping capability. I am planning to improve this in one of the nearest updates (possibly Q1/Q2 2022) by integrating the SkiaSharp.HarfBuzz nuget. I have never worked with Arabic (or equally complex languages). Therefore, I don't know yet how complex this feature is. 

Would you like to help me explore this domain (requirements / use cases) and test the experimental releases? If everything goes well, it may be the February or March release 😁

**Comment by blockhunts** [comment]
Will this case also applied to chinese/japanese characters?

**Comment by blockhunts** [comment]
@MarcinZiabek Apparently it cannot properly Japanese characters, possibly will be same with Chinese characters.

![image](https://user-images.githubusercontent.com/38546608/152072315-ba9eb7dc-024d-4a61-87d0-8133c922b9c2.png)

**Comment by girlpunk** [comment]
@blockhunts Does the font you're using contain those characters? That looks like it could be the unicode replacement character getting swapped in, which suggests the Japanese codepages aren't included.

**Comment by blockhunts** [comment]
@girlpunk  ahhh, that may be the case. I just tried to follow the getting started tutorial. And haven't fiddle with the other aspects like fonts.

**Comment by MarcinZiabek** [comment]
> @blockhunts Does the font you're using contain those characters? That looks like it could be the unicode replacement character getting swapped in, which suggests the Japanese codepages aren't included.

I agree, this may be a potential reason. Unfortunately, QuestPDF does not support font fallback at this moment 😅 I am planning to cover it in one of the future updates but there are a lot of priorities to cover.

**Comment by blockhunts** [comment]
@MarcinZiabek I see, I will patiently wait then 😄 
Keep up the good works and cheers 🥂

**Comment by ehajri** [comment]
@MarcinZiabek any update on supporting Arabic language?

**Comment by MarcinZiabek** [comment]
> @MarcinZiabek any update on supporting Arabic language?

This feature is on the roadmap. In the nearest future I am planning to take care of more advanced text related capabilities, including Arabic language support and text shaping. In fact, this is my highest priority at this moment.

Lately, a lot is happening in my personal life (positive yet time-consuminn events) and some plans become slightly delayed.

**Comment by mustafamagdy** [comment]
@MarcinZiabek Any plans when Arabic support will be available? can you guide us to do a workaround if it will take time for you to add Arabic support natively in the library?. Much appreciated the great work.

**Comment by mustafamagdy** [comment]
@megamanhxh  Thanks a lot. That is a good workaround. I also found [simpler one,](https://github.com/ibrhabash/AraibcPdfUnicodeGlyphsResharper/blob/master/AraibcPdfExtention.cs) I had to test it, though.

**Comment by GeeSuth** [comment]
@megamanhxh @mustafamagdy 
there are a clear different between twice 
`
|                 Method |      Mean |     Error |    StdDev |  Gen 0 | Allocated |
|----------------------- |----------:|----------:|----------:|-------:|----------:|
| [AraibcPdfExtentionName ](https://raw.githubusercontent.com/ibrhabash/AraibcPdfUnicodeGlyphsResharper/master/AraibcPdfExtention.cs)| 17.448 us | 0.8517 us | 2.4710 us | 6.5002 |     13 KB |
|        [ArabicFixerName ](https://raw.githubusercontent.com/Konash/arabic-support-unity/master/Assets/ArabicSupport/Scripts/ArabicSupport.cs)|  7.483 us | 0.2679 us | 0.7729 us | 2.0905 |      4 KB |
`  

tested for "بسم الله الرحمن الرحيم"


@MarcinZiabek  I don't know if that possible to combine best class to your Library ?

**Comment by MarcinZiabek** [comment]
@mustafamagdy @megamanhxh @GeeSuth Great news! I have just finished integrating the text-shaping algorithm in QuestPDF. It uses the HarfBuzz library and supports many different languages, including Arabic of course, all depends on the font used. I think this is much more stable and reliable solution, with a very little performance cost.

Please download the QuestPDF 2022.6.0-prerelease, test it in your solutions and let me know if everything is all right 😁

Also, I am hoping to take care of the RTL mode in the next release. I hope that we can cooperate on it. I have a general idea about this mode, not sure about all details involved.

**Comment by GeeSuth** [comment]
@MarcinZiabek I think ``QuestPDF 2022.6.0-prerelease `` is not uploaded yet ? 
I searching about it in tag but not found!

**Comment by MarcinZiabek** [comment]
Make sure that your IDE searches also packages with the prerelease tag. For example, there a special checkbox in Visual Studio 😁

**Comment by GeeSuth** [comment]
![image](https://user-images.githubusercontent.com/10328974/171264278-a2ce7bdf-9984-4c88-a680-92d9f46d9939.png)


Great Job 👍

**Comment by mseada94** [comment]
I find this package which still in development that support better text rendering with SkiaSharp
https://github.com/toptensoftware/RichTextKit

**Comment by MarcinZiabek** [comment]
> I find this package which still in development that support better text rendering with SkiaSharp
> https://github.com/toptensoftware/RichTextKit

You are correct, the RichTextKit library handles text in a superior way. That being said, improved text handling is one of the most important areas of improvement in the current QuestPDF roadmap. I hope that the recently introduced dual-licensing model will provide some funding and allow me to focus more on those crucial things 😀

**Comment by MarcinZiabek** [comment]
The text rendering engine has been redesigned in the 2024.3 release. Its capabilities are significantly higher and cover most use cases, including support for right-to-left languages, bidirectional text, and text-shaping. Thank you for your feedback ❤  I am closing this issue.
